File: Shortest Path Problems on Graphs/Bài_toán_16.py
import heapq

def dijkstra_general(n, adj, src):
    dist = [float('inf')] * n
    dist[src] = 0
    pq = [(0, src)]

    while pq:
        du, u = heapq.heappop(pq)
        if du > dist[u]:
            continue
        for v, w in adj[u]:
            if dist[v] > du + w:
                dist[v] = du + w
                heapq.heappush(pq, (dist[v], v))

    for i in range(n):
        print(f"Khoảng cách từ {src} đến {i}: {dist[i]}")

n = 5
adj = [[] for _ in range(n)]

adj[0].append((1, 2))
adj[0].append((1, 1))
adj[0].append((0, 0))
adj[1].append((2, 4))
adj[1].append((3, 6))
adj[2].append((3, 1))
adj[3].append((4, 3))
adj[3].append((4, 2))

dijkstra_general(n, adj, 0)

File: Shortest Path Problems on Graphs/Bài_toán_16.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <climits>
using namespace std;

typedef pair<int, int> pii;

void dijkstra_general(int n, vector<vector<pii>> &adj, int src) {
    vector<int> dist(n, INT_MAX);
    dist[src] = 0;
    priority_queue<pii, vector<pii>, greater<pii>> pq;
    pq.push({0, src});

    while (!pq.empty()) {
        auto [du, u] = pq.top(); pq.pop();
        if (du > dist[u]) continue;

        for (auto [v, w] : adj[u]) {
            if (dist[v] > du + w) {
                dist[v] = du + w;
                pq.push({dist[v], v});
            }
        }
    }

    for (int i = 0; i < n; ++i)
        cout << "Khoảng cách từ " << src << " đến " << i << ": " << dist[i] << endl;
}

int main() {
    int n = 5;
    vector<vector<pii>> adj(n);

    adj[0].push_back({1, 2});
    adj[0].push_back({1, 1});
    adj[0].push_back({0, 0});

    adj[1].push_back({2, 1});
    adj[1].push_back({3, 5});

    adj[2].push_back({3, 1});
    adj[3].push_back({4, 1});
    adj[3].push_back({4, 3});

    dijkstra_general(n, adj, 0);
    return 0;
}
